share/shwap/p2p/bitswap/block_fetch.go

```diff
@@ -0,0 +1,143 @@
+package bitswap
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"sync"
+
+	"github.com/ipfs/boxo/exchange"
+	"github.com/ipfs/go-cid"
+
+	"github.com/celestiaorg/celestia-node/share"
+)
+
+// Fetch fetches and populates given Blocks using Fetcher wrapping Bitswap.
+//
+// Validates Block against the given Root and skips Blocks that are already populated.
+// Gracefully synchronize identical Blocks requested simultaneously.
+// Blocks until either context is canceled or all Blocks are fetched and populated.
+func Fetch(ctx context.Context, fetcher exchange.Fetcher, root *share.Root, blks ...Block) error {
+	cids := make([]cid.Cid, 0, len(blks))
+	duplicate := make(map[cid.Cid]Block)
+	for _, blk := range blks {
+		if !blk.IsEmpty() {
+			continue // skip populated Blocks
+		}
+
+		idStr := blk.String()
+		p := blk.Populate(root)
+		cid := blk.CID()
+		cids = append(cids, cid)
+		// store the PopulateFn s.t. hasher can access it
+		// and fill in the Block
+		_, exists := populators.LoadOrStore(idStr, p)
+		if exists {
+			// in case there is ongoing fetch happening for the same Block elsewhere
+			// and PopulateFn has already been set -- mark the Block as duplicate
+			duplicate[cid] = blk
+		} else {
+			// only do the cleanup if we stored the PopulateFn
+			defer populators.Delete(idStr)
+		}
+	}
+
+	blkCh, err := fetcher.GetBlocks(ctx, cids)
+	if err != nil {
+		return fmt.Errorf("fetching bitswap blocks: %w", err)
+	}
+
+	for blk := range blkCh { // GetBlocks closes blkCh on ctx cancellation
+		// check if the blk is a duplicate
+		id, ok := duplicate[blk.Cid()]
+		if !ok {
+			continue
+		}
+		// if it is, we have to populate it ourselves instead of hasher,
+		// as there is only one PopulateFN allowed per ID
+		err := id.Populate(root)(blk.RawData())
+		if err != nil {
+			// this case should never happen in practice
+			// and if so something is really wrong
+			panic(fmt.Sprintf("populating duplicate block: %s", err))
```

Comment:
On the other side comment explains why panic is ok here.

Reply:
We always disagree on such things. I don't think this should be an error, but panic as is. If this happens, it means there is a critical issue that may lead to other issues, and the application should not continue running. This type of panicking is idiomatic and we can see this with concurrency writes, no copy checks in concurrency primitives, etc.

Reply:
I agree on motivation of panic, but it is not applicable to this case. Populate function here perform not only population, but also validation of data of incoming response. I If remote node will send incorrect block(response), this panic will crash our client. I think it is not acceptable to handle invalid responses with panic and panic introduces critical issue here.

Side note: Validate was even called Validate previously. Perhaps should be called as ValidateAndPopulate instead, so it is more obvious that it is totally fine to return errors

Reply:
```
func (rb *RowBlock) Populate(root *share.Root) PopulateFn {
	return func(data []byte) error {
		if !rb.IsEmpty() {
			return nil
		}
		var rowBlk bitswappb.RowBlock
		if err := rowBlk.Unmarshal(data); err != nil {
			return fmt.Errorf("unmarshaling RowBlock: %w", err)
		}

		cntr := shwap.RowFromProto(rowBlk.Row)
		if err := cntr.Validate(root, rb.ID.RowIndex); err != nil {
			return fmt.Errorf("validating Row: %w", err)
		}
		rb.Container = &cntr

		// NOTE: We don't have to validate Block in the RowBlock, as it's implicitly verified by string
		// equality of globalVerifiers entry key(requesting side) and hasher accessing the entry(response
		// verification)
		return nil
	}
}
```

If Validate() fails here, it will return error that will be thrown right into the panic

Reply:
That's the point. It should never happen if it previously didn't error with populate